lib/src/front_end/piece_factory.dart

```diff
@@ -219,6 +219,142 @@ mixin PieceFactory {
     });
   }
 
+  /// Creates a piece for the parentheses and inner parts of a for statement or
+  /// for element.
+  Piece createForLoopParts(Token leftParenthesis, ForLoopParts forLoopParts,
+      Token rightParenthesis) {
+    switch (forLoopParts) {
+      // Edge case: A totally empty for loop is formatted just as `(;;)` with
+      // no splits or spaces anywhere.
+      case ForPartsWithExpression(
+                initialization: null,
+                leftSeparator: Token(precedingComments: null),
+                condition: null,
+                rightSeparator: Token(precedingComments: null),
+                updaters: NodeList(isEmpty: true),
+              ) &&
+              var forParts
+          when rightParenthesis.precedingComments == null:
+        return buildPiece((b) {
+          b.token(leftParenthesis);
+          b.token(forParts.leftSeparator);
+          b.token(forParts.rightSeparator);
```

Comment:
```suggestion
      case ForPartsWithExpression(
            initialization: null,
            leftSeparator: Token(precedingComments: null),
            condition: null,
            rightSeparator: Token(precedingComments: null),
            updaters: NodeList(isEmpty: true),
          )
          when rightParenthesis.precedingComments == null:
        return buildPiece((b) {
          b.token(leftParenthesis);
          b.token(forLoopParts.leftSeparator);
          b.token(forLoopParts.rightSeparator);
```

I don't think we should need `&& var forParts` here. The `forLoopParts` variable should be promoted right?

Reply:
Good catch! Fixed locally.